src/librustc_mir/util/pretty.rs

```diff
@@ -534,6 +538,201 @@ pub fn write_mir_intro<'tcx>(
     Ok(())
 }
 
+/// Find all `AllocId`s mentioned (recursively) in the MIR body and print their corresponding
+/// allocations.
+pub fn write_allocations<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    body: &Body<'_>,
+    w: &mut dyn Write,
+) -> io::Result<()> {
+    fn alloc_ids_from_alloc(alloc: &Allocation) -> impl Iterator<Item = AllocId> + '_ {
+        alloc.relocations().iter().map(|(_, (_, id))| *id)
+    }
+    fn alloc_ids_from_const(val: ConstValue<'_>, mut f: impl FnMut(AllocId)) {
+        match val {
+            ConstValue::Scalar(interpret::Scalar::Ptr(ptr)) => {
+                f(ptr.alloc_id);
+            }
+            ConstValue::Scalar(interpret::Scalar::Raw { .. }) => {}
+            ConstValue::ByRef { alloc, .. } | ConstValue::Slice { data: alloc, .. } => {
+                for id in alloc_ids_from_alloc(alloc) {
+                    f(id);
+                }
+            }
+        }
+    }
+    struct CollectAllocIds(FxHashSet<AllocId>);
+    impl Visitor<'_> for CollectAllocIds {
+        fn visit_constant(&mut self, constant: &Constant<'tcx>, location: Location) {
+            self.super_constant(constant, location);
+            let val = match constant.literal.val {
+                ty::ConstKind::Value(val) => val,
+                _ => return,
+            };
+            alloc_ids_from_const(val, |id| {
+                self.0.insert(id);
+            });
+        }
+    }
+    let mut visitor = CollectAllocIds(Default::default());
+    visitor.visit_body(body);
+    let mut seen = visitor.0;
+    let mut todo: Vec<_> = seen.iter().copied().collect();
+    while let Some(id) = todo.pop() {
+        let mut write_header_and_allocation =
+            |w: &mut dyn Write, alloc: &Allocation| -> io::Result<()> {
+                write!(w, "size: {}, align: {})", alloc.size.bytes(), alloc.align.bytes())?;
+                if alloc.size == Size::ZERO {
+                    write!(w, " {{}}")?;
+                } else {
+                    writeln!(w, " {{")?;
+                    write_allocation(tcx, alloc, w, "    ")?;
+                    write!(w, "}}")?;
+                    for id in alloc_ids_from_alloc(alloc) {
```

Comment:
Easiest way to make this show up correctly is to add a `.rev()` here.
Otherwise, you're popping them in the opposite order that they show up in the output.